abstract_replication_strategy: avoid reactor stalls in `get_address_ranges` and friends

kbr- opened the pull request:
The algorithm used in `get_address_ranges` and `get_range_addresses`
calls `calculate_natural_endpoints` in a loop; the loop iterates over
all tokens in the token ring. If the complexity of a particular
implementation of `calculate_natural_endpoints` is large - say `θ(n)`,
where `n` is the number of tokens - this results in an `θ(n^2)`
algorithm (or worse). This case happens for `Everywhere` replication strategy.

For small clusters this doesn't matter that much, but if `n` is, say, `20*255`,
this may result in huge reactor stalls, as observed in practice.

We avoid these stalls by inserting tactical yields. We hope that
some day someone actually implements a subquadratic algortihm here.

The commit also adds a comment on
`abstract_replication_strategy::calculate_natural_endpoints` explaining
that the interface does not give a complexity guarantee (at this point);
the different implementations have different complexities.

For example, `Everywhere` implementation always iterates over all tokens
in the token ring, so it has `θ(n)` worst and best case complexity.
On the other hand, `NetworkTopologyStrategy` implementation usually
finishes after visiting a small part of the token ring (specifically,
as soon as it finds a token for each node in the ring) and performs
a constant number of operations for each visited token on average,
but theoretically its worst case complexity is actually `O(n + k^2)`,
where `n` is the number of all tokens and `k` is the number of endpoints
(the `k^2` appears since for each endpoint we must perform finds and
inserts on `unordered_set` of size `O(k)`; `unordered_set` operations
have `O(1)` average complexity but `O(size of the set)` worst case
complexity).

Therefore it's not easy to put any complexity guarantee in the interface
at this point. Instead, we say that:
- some implementations may yield - if their complexities force us to do so
- but in general, there is no guarantee that the implementation may
  yield - e.g. the `Everywhere` implementation does not yield.

Fixes #8555.

haaawk reviewed (approved):
LGTM

bhalevy commented on locator/abstract_replication_strategy.cc:
```diff
@@ -235,14 +235,20 @@ std::unordered_multimap<inet_address, dht::token_range>
+        // Each iteration of this outer for loop has O(number of tokens in the ring) complexity so it's fast,
+        // but in total the loop has O((number of tokens in the ring)^2) complexity; yield after each iteration to avoid stalls.
+        if (can_yield) {
+            seastar::thread::maybe_yield();
```
We already pass `can_yield` to `calculate_natural_endpoints`.
`network_topology_strategy::calculate_natural_endpoints`, for example, considers yielding based on that (See ba31350239), and so are the other implementations.
Why not add this just to `everywhere_replication_strategy::calculate_natural_endpoints`?

kbr- replied:
@asias @bhalevy 
`calculate_natural_endpoints` is a fast O(n) algorithm, my reasoning was that there is no point in yielding in the middle of computations that are almost instantaneous.
I could add a yield at the beginning of the function like this:
```
inet_address_vector_replica_set everywhere_replication_strategy::calculate_natural_endpoints(const token& search_token, const token_metadata& tm, can_yield) const {
    if (can_yield) {
        seastar::thread::maybe_yield();
    }
    return boost::copy_range<inet_address_vector_replica_set>(tm.get_all_endpoints());
}
```
but this seems a bit hacky. And I don't see much point in yielding in `get_all_endpoints` too for the same reason (it's fast). I think `calculate_natural_endpoints` could easily be a non-async function which doesn't take `can_yield` parameter due to its speed. It's the functions which call this repeatedly in a loop that cause the problem, and these functions should be the ones yielding IMO. I.e.: you don't cause a problem (you don't perform a long computation) - you don't yield; you cause a problem, you must yield.

bhalevy replied:
It's the current design.
Maybe it isn't ideal, but it's better to be consistent with it or change it system-wide, not do a partial change just for this case.

avikivity replied:
No point in yielding in the beginning, it's the responsibility of the caller in case the quota was already exceeded.

Agree that O(n) algorithms (for reasonable n) do not need yields.

kbr- replied:
So I'm not sure what to do to make every reviewer happy.
>  Maybe it isn't ideal, but it's better to be consistent with it or change it system-wide, not do a partial change just for this case.

Well I'm not changing/removing existing yields, I'm just adding new yields on top.

Here are the options I see:
1. Yield at beginning/end of `everywhere_replication_strategy::calculate_natural_endpoints` <- but see Avi's concern (which I agree with)
2. Or modify `everywhere_replication_strategy::calculate_natural_endpoints` by expanding the loop that copies endpoints from `tm` and putting the yield inside the expanded loop <- but that's just obfuscating the logic with disputable gain, I much prefer the `boost::copy_range`
3. Or leave it as it is <- I'm in favor of this one; it seems that @bhalevy does not disagree with the reasoning that there is no point yielding inside fast algorithms, the only concern here is that it "mixes" two approaches. However, maybe it's good that we introduce this new approach into the code right now, as it will pressure us to continue going into that direction in later changes. If we keep doing things the old way then the old way will prevail.

kbr- replied:
I guess there's also option
4. system-wide refactor so we go with the new approach (never yield inside fast algorithms - only functions which "really" introduce a problem must yield) <- yeah, but I don't see the value in doing this right now, the situation may change. I'd prefer fixing this bug with an easy patch right now. If we decide to do the refactor this PR would make the work a bit smaller so it's a step in the right direction nevertheless.

avikivity replied:
What about

5. Agree that yielding should be done by slow functions, not fast functions, but do not require an overnight change.

kbr- replied:
I think that would correspond to option 3.

avikivity replied:
Practically, yes, but it also includes a statement that the status quo is wrong and we want to change it.

kbr- replied:
And I guess you'd like this statement to be included in one of the comments somewhere (like a FIXME) or/and commit message?

Also, cc @bhalevy @asias - what do you think? You suggested to put yields in `calculate_natural_endpoints` and `token_metadata::get_all_endpoints` as well (probably in a followup), but if we go with option 3./5. discussed above, this would actually mean we go in a completely opposite direction - in particular, _remove_ `can_yield` parameter from `calculate_natural_endpoints` (because it's a linear algorithm... although I'm not 100% sure for NTS implementation, but it _should_ be linear, I know from a high level that it _can_ be) (removing `can_yield` from there could be done eventually, not necessarily in this PR). Are you fine with that?

bhalevy replied:
I'm certainly okay with making forward progress and not sticking mistakes we made in the past.
Meaning we can go ahead with the approach in this patch, documenting (with FIXME comment(s)), and opening github issue(s) to fix them and remove yielding from the lower level functions that have linear complexity.

asias replied:
The n for calculate_natural_endpoints is number of nodes * number of tokens per node. This can be really big in large clusters, especially if user uses a much bigger  num_tokens (default is 256).  I am not so confident it is 100% stall free. 

To move forward, I am fine with this patch, since we use get_address_ranges in limited places.

asias commented:
Hello Kamil, Thanks! Some comments and ideas below:

- The commit message misses which issue it fixes.

- I think we should fix everywhere_replication_strategy::calculate_natural_endpoints to respect can_yield instead of yield in get_address_ranges.

- We can yield in tm::get_all_endpoints too.

- As a follow up, we can even introduce a filed inside token_metadata to track endpoints directly  (modified whenever we modify _token_to_endpoint_map). This avoids walking through the map to get the endpoints. And the count_normal_token_owners will be O(1).

asias commented:
> Hello Kamil, Thanks! Some comments and ideas below:
> 
>     * The commit message misses which issue it fixes.
> 
>     * I think we should fix everywhere_replication_strategy::calculate_natural_endpoints to respect can_yield instead of yield in get_address_ranges.
> 
>     * We can yield in tm::get_all_endpoints too.
> 
>     * As a follow up, we can even introduce a filed inside token_metadata to track endpoints directly  (modified whenever we modify _token_to_endpoint_map). This avoids walking through the map to get the endpoints. And the count_normal_token_owners will be O(1).

I created https://github.com/scylladb/scylla/issues/8673 to track third item above.

kbr- commented:
So after more thinking I came to the conclusion that removing `can_yield` from `calculate_natural_endpoints` is not a good idea.

The problem is that `abstract_replication_strategy::calculate_natural_endpoints` is virtual - it's an abstract interface - and there is no way to put a global complexity guarantee on all implementations, at least at this point. Well, we could, but it would be quite large, suggesting that the implementation should yield. For example, I analyzed the implementation for `network_topology_strategy` and found that, if I'm not mistaken:
- the average complexity is `O(k)`, where `k` is the number of endpoints - and this is what happens in practice
- but, considering that in the worst case we may need to visit the entire token ring (although this probably never happens), and that `unordered_set` operations in the worst case have linear complexity (in the size of the set; again, probably never happens), the worst case complexity is `O(n + k^2)`, where `n = number of all tokens` and `k = number of endpoints`.

So considering the worst case complexity, and considering that `unordered_set` operations are quite expensive even when they are constant on average - `network_topology_strategy` should probably yield.

However it doesn't make sense for all implementations to yield. For example, it doesn't make much sense for `everywhere_replication_strategy` to yield because it's just a `O(n)` copy of a vector.

Thus in the new version of the PR I added a comment at the `calculate_natural_endpoints` interface that the implementation doesn't guarantee a yield - it will only yield if it's internally necessary, hence the caller is still responsible for yielding if they call the function in a loop. I also removed some comments about complexity added in the previous version that were imprecise/incorrect.

I also explained the problem in commit message and updated the PR cover letter.

bhalevy reviewed (approved):
LGTM

kbr- commented:
@avikivity looks like this is ready to merge, if there are no further objections.

avikivity commented:
No further objections, your honor.
